File: python/2022/day02.py
SHAPES = {
    "A": "rock",
    "B": "paper",
    "C": "scissors",
    "X": "rock",
    "Y": "paper",
    "Z": "scissors",
}

SHAPE_SCORES = {
    "rock": 1,
    "paper": 2,
    "scissors": 3,
}

SHAPE_DEFEATS = {
    "rock": "scissors",
    "paper": "rock",
    "scissors": "paper",
}

OUTCOMES = {
    "X": "lose",
    "Y": "draw",
    "Z": "win",
}


def part1(input: str) -> int:
    lines = input.splitlines()
    score = 0

    for line in lines:
        [opp, me] = line.split()
        opp = SHAPES[opp]
        me = SHAPES[me]
        score += SHAPE_SCORES[me]

        if opp == me:
            score += 3

        if SHAPE_DEFEATS[me] == opp:
            score += 6

    return score


def part2(input: str) -> int:
    lines = input.splitlines()
    score = 0

    for line in lines:
        [opp, me] = line.split()
        opp = SHAPES[opp]
        outcome = OUTCOMES[me]

        match outcome:
            case "lose":
                me = SHAPE_DEFEATS[opp]

            case "draw":
                score += 3
                me = opp

            case "win":
                score += 6
                me = SHAPE_DEFEATS[SHAPE_DEFEATS[opp]]

            case _:
                pass

        score += SHAPE_SCORES[me]

    return score


TEST_INPUT = """A Y
B X
C Z"""

PART1_TESTS = [
    (TEST_INPUT, 15)
]

PART2_TESTS = [
    (TEST_INPUT, 12)
]

File: rust/2022/src/day02.rs
enum Moves {
    Rock,
    Paper,
    Scissors
}

#[allow(dead_code)]
pub fn puzzle() {
    let input = crate::input::load_input(2);
    let mut score = 0;
    let mut opp_move: Moves;
    let mut my_move: Moves;

    for line in input.lines() {
        match line.chars().next().unwrap() {
            'A' => opp_move = Moves::Rock,
            'B' => opp_move = Moves::Paper,
            'C' => opp_move = Moves::Scissors,
            _ => panic!("failed to match opp_move"),
        }

        match line.chars().nth(2).unwrap() {
            'X' => my_move = Moves::Rock,
            'Y' => my_move = Moves::Paper,
            'Z' => my_move = Moves::Scissors,
            _ => panic!("failed to match my_move"),
        }

        match my_move {
            Moves::Rock => {
                match opp_move {
                    Moves::Rock => score += 4,
                    Moves::Paper => score += 1,
                    Moves::Scissors => score += 7,
                }
            },
            Moves::Paper => {
                match opp_move {
                    Moves::Rock => score += 8,
                    Moves::Paper => score += 5,
                    Moves::Scissors => score += 2,
                }
            },
            Moves::Scissors => {
                match opp_move {
                    Moves::Rock => score += 3,
                    Moves::Paper => score += 9,
                    Moves::Scissors => score += 6,
                }
            },
        }
    }

    println!("Part 1 answer: {score}");

    score = 0;

    for line in input.lines() {
        match line.chars().next().unwrap() {
            'A' => opp_move = Moves::Rock,
            'B' => opp_move = Moves::Paper,
            'C' => opp_move = Moves::Scissors,
            _ => panic!("failed to match opp_move"),
        }

        match line.chars().nth(2).unwrap() {
            'X' =>  {
                match opp_move {
                    Moves::Rock => score += 3,
                    Moves::Paper => score += 1,
                    Moves::Scissors => score += 2,
                }
            },
            'Y' => {
                match opp_move {
                    Moves::Rock => score += 4,
                    Moves::Paper => score += 5,
                    Moves::Scissors => score += 6,
                }
            },
            'Z' => {
                match opp_move {
                    Moves::Rock => score += 8,
                    Moves::Paper => score += 9,
                    Moves::Scissors => score += 7,
                }
            },
            _ => panic!("failed to match my_move"),
        }
    }
    
    println!("Part 2 answer: {score}");
}
